ls1intum/Artemis
# `Exam mode`: Add progress monitoring for exercise start preparation

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



The "preparse exercise start" action that can be applied on all student exams to create participations before the exercise starts can take a long time for large courses. Right now, it's implemented so that the HTTP request doesn't return until it's finished and it's likely that this takes longer than the HTTP timeout. Also, the user does not have any idea how long this will take, and how much is already done.

### Description


This PR adds a progress bar that updates live via the websocket to the student exams view that indicates the progress of the exercise preparation process, including a calculated estimation of the remaining time.

The initial request to run this process will only start it and then return. The status is distributed to everyone who is currently on that student exam page (it has authorization checks in place as well). Additionally, the current status is stored in a cache so if a new user joins the site or the current user reloads, the status can be fetched and displayed immediately without waiting for the next WS update as otherwise the user could think that the process is not running; additionally, the cache allows to show a completed process for a while after it finished as well so the users know there was one in the short past.

### Steps for Testing

Prerequisites:
- 1-2 Instructor(s)
- 1 Exam with a lot of registered students so the preparation takes long enough to observe it

1. Go to the student exams view of the exam
2. If not already done, generate student exams
3. Start the exercise preparation process by clicking the correct button
4. Observe the process in the progress bar. 
5. Verify that it survives reloads, and that it displays for another instructor on the same page as well

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| student-exams.component.ts | 73% | 93% |
| exam-management.service.ts | 73% | 94% |
| ExamService.java | 82% | 96% |
| StudentExamService.java | 76% | 88% |
| StudentExamResource.java | 77% | 92% |
| WebsocketMessagingService.java | 100% | 95% |
| WebsocketConfiguration.TopicSubscriptionInterceptor | 68% | 75% |

### Screenshots



https://user-images.githubusercontent.com/23171488/181344452-33b7fe9f-750b-43b5-997a-bf97f183b91a.mov
